[Syed Muhammad Aljunied] Duke Increments

muhammadaljunied opened the pull request.

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

j-lum committed:
Change file mode on `gradle` to be executable

j-lum committed:
Merge pull request #12 from j-lum/gradle+x

Change file mode on `gradle` to be executable (#9)

j-lum committed:
Add configuration for console applications

Gradle defaults to an empty stdin which results in runtime exceptions
when attempting to read from `System.in`. Let's add some sensible
defaults for students who may still need to work with the standard
input stream.

j-lum committed:
Merge pull request #13 from j-lum/javaexec

Add configuration for console applications

Commit:
Merge branch 'gradle'

Commit:
Restructured project package to match seedu.duke
Added Gradle Support
Added Checkstyle plugin
Added Shadow plugin
Added JUnitTest Dependency

Commit:
Revert "Restructured project package to match seedu.duke"

This reverts commit 721456ddbb8b18afd97db484e0b5e3b0c0c34ff3.

Commit:
Restructured project package to match seedu.duke
Added Gradle Support
Added Checkstyle plugin
Added Shadow plugin
Added JUnitTest Dependency

Commit:
Added Text UI Test script

Commit:
Correct directory structure for main.

Commit:
Added Greeting
Added Echo
Added Exit

Commit:
Implemented "Add to list" feature
Implemented "List all task" feature

Commit:
Resolve Conflicts

Commit:
Revert "Implemented "Add to list" feature"

This reverts commit 708252ae7a19018d4b7d6e2316c9f867fa1770e0.

Commit:
Implemented "Add to list" feature
Implemented "List all task" feature

Commit:
Replaced Horizontal Line printing with Method printLine()

Commit:
Added Enums for Commands to easily identify.
Implemented pattern matching for commands and arguments with regex
Abstracted out Commands to reduce excessive if-else for executing commands
Specific commands print their relevant messages
Added Mark as completed

Commit:
Added error handling for parsing of command input
Resolved Checkstyle Errors
Added package seedu.duke.trackables ahead of level-4 requirements

Commit:
Added Abstraction for Tasks - Event, Todo & Deadline
Refactored task object for abstraction
Refactored Command objects to reflect abstraction of task object.

Commit:
Added Commands to support 3 new trackables

Commit:
Resolved CheckStyle Issues
Updated patterns for Regex
Added Support for Event, Todo, Deadline
Added Error Abstractions for Common errors (Invalid Command, Invalid Arguments)
Duke specific exceptions are thrown with original cause (if available) to provide more useful stack trace.

** Forgot to commit Level 4, thus tagged as level 4 and 5)

Commit:
Delete feature added

TSAI-HSIAO-HAN commented on src/main/java/seedu/duke/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,27 @@
+    @Override
+    public void execute(List<Task> tasks) {
+        tasks.add(deadline);
+        echo(
```
The indentation for wrapped lines in execute method should be 8 spaces.

muhammadaljunied replied:
Line 20 should be together with Line 19.

TSAI-HSIAO-HAN commented on src/main/java/seedu/duke/commands/DoneCommand.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public void execute(List<Task> tasks) {
+        tasks.get(taskId).markAsDone();
+        echo(
```
The indentation for wrapped lines in execute method should be 8 spaces.

muhammadaljunied replied:
Line 19 should be together with Line 18.

TSAI-HSIAO-HAN commented on src/main/java/seedu/duke/commands/EventCommand.java:
```diff
@@ -0,0 +1,25 @@
+    @Override
+    public void execute(List<Task> tasks) {
+        tasks.add(event);
+        echo(new String[] {
```
The indentation for wrapped lines in execute method should be 8 spaces.

muhammadaljunied replied:
This is the indentation that CheckStyle approves. 

Line 19 at level 8
Line 20-22 at level 12
Line 23 at level 8 (to close the String array instantiation)

TSAI-HSIAO-HAN reviewed (commented):
The overall Java style seems correct. Can break some lines to increase readability.

Commit:
Resolve Formatting issue for PR Comment

ChrisKheng commented:
**Comment**
1. 
* Regarding **Duke** class, maybe you want to separate the functionalities in it into more classes? Like **Storage**, **Parser** etc. I think there are too many functionalities in the **Duke** class and it doesn't really correspond to the Single Responsibility principle in SOLID. It may make the process of changing the program's code to fit new requirement harder as it may be quite hard to locate the right piece of code to change among the extensive lines of code. 

* Also, I think the purpose of Duke class should act like a **body** which links all its parts(Storage, Parser, TaskList) together and not doing too much of logic. This concept may help you to refactor your code.

2. 
* Similar to 1., maybe you want to separate the logic of handling database, printing into more classes such as **Storage** and **Parser**? This would reduce the responsibility of the **Command** classes and may aid in coding or making changes when doing further extensions of the program.

* In the **DeleteCommand** class, maybe you want to consider not to decrease the index of the task by 1? Cuz I think maybe in the future you may need to retrieve the index of the task from the command and hence it may cause confusion or bug when the index is different from the "surface" index (i.e. the id starts from 1 when displaying the tasks). You can decrease the index by 1 in the execute() method when trying to retrieve the task.

* In the **DoneCommand**, maybe you want to assign the task that you have retrieved to a variable so that later you can directly call toString() on it? This will then prevent from doing two get(), which means traversing twice through the list, which can increase the complexity of the solution.

* Also, it may be good for you to rename the **TodoCommand**, **DeadlineCommand** and **EventCommand** to maybe **AddToDoCommand**, **AddDeadlineCommand** and **AddEventCommand**? This can improve tbe clarity of the code.

3. 
* Regarding the **Task** classes, maybe it would be good to make an enum to label the respective classes (ToDo, Deadline, Event)? This may help in the future when you want to extend your program.

* Regarding **Event** class, maybe you want to use another name such as "time" instead of "at" for the string which stores the date? I think this would improve the clarity of your code. The same thing applies to **Deadline** class too, maybe you want to use "deadline" instead of "by"?

Overall, I think you could improve on the class structure of the program to make it easier to manage the program, especially when the program scale becomes larger and larger. All the best and cheers!

ChrisKheng commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,188 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Duke {
```
Regarding Duke class, maybe you want to separate the functionalities in it into more classes? Like Storage, Parser etc. I think there are too many functionalities in the Duke class and it doesn't really correspond to the Single Responsibility principle in SOLID. It may make the process of changing the program's code to fit new requirement harder as it may be quite hard to locate the right piece of code to change among the extensive lines of code.

Also, I think the purpose of Duke class should act like a body which links all its parts(Storage, Parser, TaskList) together and not doing too much of logic. This concept may help you to refactor your code.

ChrisKheng commented on src/main/java/seedu/duke/commands/Command.java:
```diff
@@ -0,0 +1,35 @@
+     * Prints out the message.
+     * @param messages Array of messages to show.
+     */
+    protected void echo(String[] messages) {
```
Maybe you want to separate the logic of printing, handling database into more classes such as Storage and Parser? This would reduce the responsibility of the Command classes and may aid in coding or making changes when doing further extensions of the program.

ChrisKheng commented on src/main/java/seedu/duke/commands/DoneCommand.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public void execute(List<Task> tasks) {
+        tasks.get(taskId).markAsDone();
```
Maybe you want to assign the task that you have retrieved to a variable so that later you can directly call toString() on it? This will then prevent from doing two get(), which means traversing twice through the list, which can increase the complexity of the solution.

ChrisKheng commented on src/main/java/seedu/duke/commands/TodoCommand.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.List;
+
+public class TodoCommand extends Command {
```
Maybe it's good to rename the TodoCommand, DeadlineCommand and EventCommand to maybe AddToDoCommand, AddDeadlineCommand and AddEventCommand? This can improve the clarity of the code.

ChrisKheng commented on src/main/java/seedu/duke/trackables/Task.java:
```diff
@@ -0,0 +1,28 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
+
```
Maybe it would be good to make an enum to label the respective classes (ToDo, Deadline, Event)? This may help in the future when you want to extend your program.

ChrisKheng commented on src/main/java/seedu/duke/trackables/Event.java:
```diff
@@ -0,0 +1,16 @@
+
+public class Event extends Task {
+
+    protected String at;
```
Maybe you want to use another name such as "time" instead of "at" for the string which stores the date? I think this would improve the clarity of your code.

ChrisKheng commented on src/main/java/seedu/duke/trackables/Deadline.java:
```diff
@@ -0,0 +1,16 @@
+
+public class Deadline extends Task {
+
+    protected String by;
```
Maybe you want to use "deadline" instead of "by" to improve the clarity of the code?

ChrisKheng reviewed (commented):
Last note: In the DeleteCommand class, maybe you want to consider not to decrease the index of the task by 1? Cuz I think maybe in the future you may need to retrieve the index of the task from the command and hence it may cause confusion or bug when the index is different from the "surface" index (i.e. the id starts from 1 when displaying the tasks). You can decrease the index by 1 in the execute() method when trying to retrieve the task.

Overall, I think you could improve on the structure of the program by separating the logic of the program into more classes (Single Responsibility Principle). This can make it easier to manage the program's code, especially when the program scale becomes larger and larger. All the best and cheers!

Commit:
Travis CI Configuration

Commit:
Corrected Travis Script

Commit:
Added Save/Load Feature
Resolved Checkstyle Issues

Commit:
Added support for Date (Event and Deadline type tasks)
Refactored Command Arguments parsing into respective command objects
Error handling now cleaner with more specific error messages.

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/seedu/duke/Duke.java
#	src/main/java/seedu/duke/trackables/Deadline.java

Commit:
Resolved Regression due to Level-7/Level-8 Merge and missed Merge Conflict on Duke.java.

Commit:
Abstracted out Parsing into Parser.java from Duke.java
Abstracted out Ui and printing operations to Ui.java from Duke.java
Updated Exceptions with new constructor e*(String message)

Commit:
Resolved Regressions from Abstracting UI

Commit:
Resolved regression caused by abstracting List<Task> to TaskList.java.
Updated Javadocs.
Updated UI to handle reading inputs.
Refactored IndexOutOfBounds error handling to TaskList from DeleteCommand and DoneCommand.

Commit:
Added JUnit Testing for DeleteCommand, DoneCommand and Parser

Commit:
Implemented Find functionality

Commit:
Implemented Find functionality

Commit:
Revert "Implemented Find functionality"

This reverts commit da3854ee320412423f7f51ba8a8225ced7466e88.

Commit:
Added Complete JavaDoc.

Commit:
Updated codebase to match Java Coding Standards.

Commit:
Merge branch 'branch-JavaDoc'

* branch-JavaDoc:
  Added Complete JavaDoc.

Commit:
Merge branch 'branch-CodingStandard'

* branch-CodingStandard:
  Updated codebase to match Java Coding Standards.

# Conflicts:
#	src/main/java/seedu/duke/commands/AddCommand.java
#	src/main/java/seedu/duke/trackables/Event.java

Commit:
Corrected the 0-Index issue with delete and done command.

Commit:
Fixed a Dead format issue and reversed the 0-index correction from the previous commit.

Commit:
Added JavaFX Support

Commit:
Updated Commands to return the response as a String.

Commit:
Resolved clipping issue with DialogBox.
Linked UI with Duke.

muhammadaljunied committed:
Added Assertions (#6)

Commit:
Updated codebase to match Code Quality.
Added missing Javadocs.
Added UI tweaks

Commit:
Updated Photos
Added Stop handler when user closes using window close
Added wrapping width for text node in UI

Commit:
Added Ui Screenshot

Commit:
Updated build for Cross Platform support.

Commit:
Incremented Build

Commit:
Updated README.md

Commit:
update README.MD

muhammadaljunied committed:
Set theme jekyll-theme-cayman
